MergerLogic/DataTypes/Gpkg.cs

```diff
@@ -73,26 +74,48 @@ public override void Reset()
             this._offset = 0;
         }
 
-        public override List<Tile> GetNextBatch(out string batchIdentifier)
+        public override List<Tile> GetNextBatch(out string currentBatchIdentifier, out string? nextBatchIdentifier, string? inCompletedBatchIdentifier, long? totalTilesCount)
         {
-            batchIdentifier = this._offset.ToString();
-            //TODO: optimize after IOC refactoring
-            int counter = 0;
-            List<Tile> tiles = this.Utils.GetBatch(this.BatchSize, this._offset)
-                .Select(t =>
+            lock (_locker)
+            {
+                if (inCompletedBatchIdentifier is not null)
+                {
+                    long.TryParse(inCompletedBatchIdentifier, out long result);
+                    this._offset = result;
+                }
+                currentBatchIdentifier = this._offset.ToString();
+                List<Tile> tiles = new List<Tile>();
+                if (this._offset != totalTilesCount)
                 {
-                    Tile tile = this.ConvertOriginTile(t);
-                    tile = this.ToCurrentGrid(tile);
-                    counter++;
-                    return tile;
-                }).Where(t => t != null).ToList();
-            this._offset += counter;
-            return tiles;
+                    //TODO: optimize after IOC refactoring
+                    int counter = 0;
+
+                    tiles = this.Utils.GetBatch(this.BatchSize, this._offset)
+                        .Select(t =>
+                        {
+                            Tile tile = this.ConvertOriginTile(t);
+                            tile = this.ToCurrentGrid(tile);
+                            counter++;
+                            return tile;
+                        }).Where(t => t != null).ToList();
+                    if (inCompletedBatchIdentifier is null)
+                    {
+                        Interlocked.Add(ref this._offset, counter);
```

Comment:
Why use `Interlocked` if the whole logic is locked?
In addition, this should always happen because the `if` statement should not exist.